File: Libraries/graphics/button/button.cpp
#include "button.hpp"

bool gfx::Button::isHovered(int mouse_x, int mouse_y) const {
    RectShape rect = getTranslatedRect();
    return !disabled && mouse_x >= rect.x && mouse_y >= rect.y && mouse_x <= rect.x + rect.w && mouse_y <= rect.y + rect.h;
}

void gfx::Button::render(int mouse_x, int mouse_y, bool button_pressed) {
    RectShape rect = getTranslatedRect();
    
    while(timer_counter < timer.getTimeElapsed()) {
        int hover_progress_target = isHovered(mouse_x, mouse_y) ? (button_pressed ? 200 : 255) : 0;
        hover_progress += float(hover_progress_target - hover_progress) / 40;
        if(std::abs(hover_progress_target - hover_progress) <= 5)
            hover_progress = hover_progress_target;
        timer_counter++;
    }
    
    Color button_color{
        (unsigned char)(hover_color.r * hover_progress / 255 + def_color.r * float(255 - hover_progress) / 255),
        (unsigned char)(hover_color.g * hover_progress / 255 + def_color.g * float(255 - hover_progress) / 255),
        (unsigned char)(hover_color.b * hover_progress / 255 + def_color.b * float(255 - hover_progress) / 255),
        (unsigned char)(hover_color.a * hover_progress / 255 + def_color.a * float(255 - hover_progress) / 255),
    };
    
    Color button_border_color{
        (unsigned char)(border_hover_color.r * hover_progress / 255 + def_border_color.r * float(255 - hover_progress) / 255),
        (unsigned char)(border_hover_color.g * hover_progress / 255 + def_border_color.g * float(255 - hover_progress) / 255),
        (unsigned char)(border_hover_color.b * hover_progress / 255 + def_border_color.b * float(255 - hover_progress) / 255),
        (unsigned char)(border_hover_color.a * hover_progress / 255 + def_border_color.a * float(255 - hover_progress) / 255),
    };
    
    int padding = (255 - hover_progress) / 255 * 30;
    RectShape hover_rect(rect.x + padding, rect.y + padding, std::max(0, rect.w - 2 * padding), std::max(0, rect.h - 2 * padding));
    rect.render(def_color);
    rect.renderOutline(def_border_color);
    hover_rect.render(button_color);
    hover_rect.renderOutline(button_border_color);
    
    float texture_scale = scale + hover_progress / 255.f * 0.2;
    Texture::render(texture_scale, rect.x + rect.w / 2 - getTextureWidth() * texture_scale / 2, rect.y + rect.h / 2 - getTextureHeight() * texture_scale / 2);
}

void gfx::Button::loadFromSurface(const Surface& surface) {
    Texture::loadFromSurface(surface);
    updateSize();
}

void gfx::Button::setScale(float scale_) {
    scale = scale_;
    updateSize();
}

float gfx::Button::getScale() const {
    return scale;
}

void gfx::Button::setMargin(int margin_) {
    margin = margin_;
    updateSize();
}

int gfx::Button::getMargin() const {
    return margin;
}

int gfx::Button::getWidth() const {
    return w;
}

int gfx::Button::getHeight() const {
    return h;
}

void gfx::Button::updateSize() {
    w = (getTextureWidth() + margin * 2) * scale;
    h = (getTextureHeight() + margin * 2) * scale;
}

File: libraries/ui/button.rs
use super::{UiContext, UiElement};
use crate::libraries::graphics as gfx;
use crate::libraries::timing;

use super::theme::{BUTTON_BORDER_COLOR, BUTTON_COLOR, BUTTON_PADDING, HOVERED_BUTTON_BORDER_COLOR, HOVERED_BUTTON_COLOR};

/// A clickable rectangle with an image in it and a hover animation.
pub struct Button {
    pub pos: gfx::FloatPos,
    pub orientation: super::Orientation,
    pub texture: gfx::Texture,
    pub padding: f32,
    pub scale: f32,
    pub color: gfx::Color,
    pub border_color: gfx::Color,
    pub hover_color: gfx::Color,
    pub hover_border_color: gfx::Color,
    pub disabled: bool,
    pub darken_on_disabled: bool,
    pub hover_progress: f32,
    /// One hover step per elapsed millisecond, bounded so that a button built long before it
    /// is first drawn - the pause menu's, say - does not owe a step for every one of them.
    animation_timer: timing::FixedStep,
    click: super::ClickTracker,
    on_click: Box<dyn Fn()>,
}

impl Button {
    #[must_use]
    pub fn new<F: 'static + Fn()>(closure: F) -> Self {
        Self {
            pos: gfx::FloatPos(0.0, 0.0),
            orientation: super::TOP_LEFT,
            texture: gfx::Texture::new(),
            padding: BUTTON_PADDING,
            scale: 1.0,
            color: BUTTON_COLOR,
            border_color: BUTTON_BORDER_COLOR,
            hover_color: HOVERED_BUTTON_COLOR,
            hover_border_color: HOVERED_BUTTON_BORDER_COLOR,
            disabled: false,
            darken_on_disabled: false,
            hover_progress: 0.0,
            animation_timer: timing::FixedStep::for_animation(1),
            click: super::ClickTracker::default(),
            on_click: Box::new(closure),
        }
    }

    /// The image plus the padding, scaled.
    #[must_use]
    pub fn get_size(&self) -> gfx::FloatSize {
        gfx::FloatSize(
            (self.texture.get_texture_size().0 + self.padding * 2.0) * self.scale,
            (self.texture.get_texture_size().1 + self.padding * 2.0) * self.scale,
        )
    }

    pub fn press(&self) {
        (self.on_click)();
    }

    /// Pins the hover animation at `progress` for the golden-image tests.
    ///
    /// The animation chases a target that depends on the real mouse position, so freezing the
    /// timer is what makes the value set here exactly what gets drawn.
    #[cfg(feature = "render-tests")]
    pub const fn settle_hover(&mut self, progress: f32) {
        self.hover_progress = progress;
        self.animation_timer.freeze();
    }
}

impl UiElement for Button {
    fn render_inner(&mut self, graphics: &mut gfx::GraphicsContext, parent_container: &super::Container) {
        let container = self.get_container(graphics, parent_container);
        let rect = container.get_absolute_rect();

        let hover_target = if self.is_hovered(graphics, parent_container) {
            if graphics.get_key_state(gfx::Key::MouseLeft) {
                0.8
            } else {
                1.0
            }
        } else {
            0.0
        };

        while self.animation_timer.step() {
            self.hover_progress = super::approach(self.hover_progress, hover_target, 40.0, 0.01);
        }

        let button_color = gfx::interpolate_colors(self.color, self.hover_color, self.hover_progress);
        let button_border_color = gfx::interpolate_colors(self.border_color, self.hover_border_color, self.hover_progress);

        // The hover fill grows out of the middle of the button as the hover comes in.
        let padding = (1.0 - self.hover_progress) * 30.0;
        let hover_rect = gfx::Rect::new(
            rect.pos + gfx::FloatPos(padding, padding),
            gfx::FloatSize(f32::max(0.0, rect.size.0 - 2.0 * padding), f32::max(0.0, rect.size.1 - 2.0 * padding)),
        );
        rect.render(graphics, self.color);
        rect.render_outline(graphics, self.border_color);
        hover_rect.render(graphics, button_color);
        hover_rect.render_outline(graphics, button_border_color);

        let texture_scale = self.scale + self.hover_progress * 0.3;
        let texture_pos = gfx::FloatPos(
            rect.pos.0 + rect.size.0 / 2.0 - self.texture.get_texture_size().0 * texture_scale / 2.0,
            rect.pos.1 + rect.size.1 / 2.0 - self.texture.get_texture_size().1 * texture_scale / 2.0,
        );
        self.texture.render(graphics, texture_scale, texture_pos, None, false, None);

        if self.disabled && self.darken_on_disabled {
            rect.render(graphics, gfx::Color::new(0, 0, 0, 100));
        }
    }

    /// Fires on the release of a press that landed on this same button - see `ClickTracker`.
    fn on_event_inner(&mut self, graphics: &mut dyn super::UiContext, event: &gfx::Event, parent_container: &super::Container) -> bool {
        let hovered = self.is_hovered(graphics, parent_container);
        if self.click.completes_a_click(event, hovered) {
            (self.on_click)();
            return true;
        }
        false
    }

    fn get_container(&self, graphics: &dyn super::UiContext, parent_container: &super::Container) -> super::Container {
        super::Container::new(graphics, self.pos, self.get_size(), self.orientation, Some(parent_container))
    }

    /// A disabled button is never hovered, which is what stops it reacting to clicks as well
    /// as what keeps it from lighting up under the pointer.
    fn is_hovered(&self, graphics: &dyn super::UiContext, parent_container: &super::Container) -> bool {
        !self.disabled && self.get_container(graphics, parent_container).get_absolute_rect().contains(graphics.get_mouse_pos())
    }
}
